text/0000-separate-context-propagation.md

```diff
@@ -0,0 +1,193 @@
+
+**ClearBaggage(context) -> context**  
+To avoid sending baggage to an untrusted downstream process, the Baggage API provides a function remove all baggage from a context, 
+
+**GetPropagator(type) -> (inject, extract)**  
+To register with the propagation system, the Baggage API provides a set of propagation functions for every propagation type.
+
+
+## Additional APIs
+
+Because the application and context propagation layers are separated, it is possible to create new distributed applications which do not depend on either the Observability or Baggage APIs.
+
+**GetPropagator(type) -> (inject, extract)**  
+To register with the propagation system, additional APIs provide a set of propagation functions for every propagation type.
+
+
+# Context Propagation Layer
+
+## Context API
+
+Distributed applications access data in-process using a shared context object. Each distributed application sets a single key in the context, containing all of the data for that system.
+
+**SetValue(context, key, value) -> context**  
+To record the local state of an application, the Context API provides a function which takes a context, a key, and a value as input, and returns an updated context which contains the new value.
+
+**GetValue(context, key) -> value**  
+To access the local state of an application, the Context API provides a function which takes a context and a key as input, and returns a value.
+
+### Optional: Automated Context Management
+When possible, context should automatically be associated with program execution . Note that some languages do not provide any facility for setting and getting a current context. In these cases, the user is responsible for managing the current context. 
+
+**SetCurrent(context)**  
+To associate a context with program execution, the Context API provides a function which takes a Context.
+
+**GetCurrent() -> context**  
+To access the context associated with program execution, the Context API provides a function which takes no arguments and returns a Context.
+
+
+## Propagation API
+
+Distributed applications send data to downstream processes via propagators, functions which read and write application context into RPC requests. Each distributed application creates a set of propagators for every type of supported medium - currently HTTP and Binary.
+
+**Inject(context, request)**  
+To send the data for all distributed applications downstream to the next process, the Propagation API provides a function which takes a context and a request, and mutates the request to include the encoded context. The canonical representation of a request is as a map.
+
+**Extract(context, request) -> context**  
+To receive data injected by prior upstream processes, the Propagation API provides a function which takes a context and a request, and returns an updated context.
```

Comment:
What does Extract update the context with? Is it SpanContext or Span?
If it is SpanContext then it would require another api to set the current SpanContext in addition to Span.
If it is Span then Extract also has to start a span which I don't think it is appropriate. 
So, it would be better for this API to return SpanContext. Unless we introduce a concept of Inactive Span which simply contains remote SpanContext.

Reply:
You are right, I also think this is an important detail that needs to be resolved.

I have another suggestion: The context always contains just a spancontext. Any SpanContext has an optional reference to the span it refers to, which is null in case of a remote span (cf https://github.com/open-telemetry/opentelemetry-specification/pull/216#issuecomment-527379659)

Reply:
I made a similar suggestion when we were working on the initial Java API in the spring. I think it has the benefits not just in the context of this discussion, but also as a alternative mechanism for implementing the "span". For example, in a streaming implementation the span does not even need to exist if all write operations simply work off the SpanContext. In other cases, spans may be kept by the tracer in a private dictionary, which makes memory management easier than if user code has an explicit reference to the span.

Reply:
I don't understand how this would reduce references to Spans. I thought the user would be able to get this reference anytime from the SpanContext. But even withot such an API, startSpan etc would of course still return references to spans.

Reply:
StartSpan would return a new span context. Span is just an accumulator of data, it is not actually needed in the API. We could move all Span methods into Tracer directly and make them accept span context as the first argument, that would be functionally equivalent.

Reply:
Ah, I understand! There is one functional difference though: if you allowed something like `Tracer.setSpanAttribute(context, "x", "y")`, you implicitly allow setting attributes on spans that were not created in this process. This is prevented by the current Span/SpanContext separation and I think that's a good thing.

Reply:
I remember that discussion on last spring, and I think the desire was that `Span`s created in the same process could share the monotonic clock (which is an optimization, more than a requirement).

Reply:
@Oberon00 indeed, you can extract a parent span context and set attributes on it. I think it's actually a feature of this approach, not the downside :-) . Although it would only work in the streaming-like implementations, not the current approach where "finished span is not writeable". But then, even today you can do `span.end(); span.setTag()`, which is meaningless.

@carlosalberto there are many ways of implementing SDK when the API only exposes Span Context. As @Oberon00 said, the simple implementation is still to have the Span at the SDK level which is linked from Span Context, this way you can share clock offsets.